### PR TITLE
Plane: Add THERMAL mode for soaring

### PR DESCRIPTION
**Pull request by samuelctabor**
This removes the long-standing hijacking of LOITER mode for use with thermal soaring. THERMAL mode is unavailable if soaring is not configured on the board, and will refuse to enter if soaring is disabled via parameter or RC switch. In future more of the functionality will be moved from soaring.cpp into mode_thermal.

~~Unfortunately I had to remove `#include "mode.h"` from the flight mode .cpp files to ensure SOARING_ENABLED in defined in mode.h. This doesn't seem ideal but it is done this way in ArduCopter.~~

Once this is merged I'll do a PR to add the new mode to pymavlink to get rid of the use of mode number in autotest.

### CODE REVIEW

File: ArduPlane/soaring.cpp [hunk 1]
```diff
@@ -26,8 +26,8 @@ void Plane::update_soaring() {
     case Mode::Number::CRUISE:
         g2.soaring_controller.suppress_throttle();
         break;
-    case Mode::Number::LOITER:
-        // Never use throttle in LOITER with soaring active.
+    case Mode::Number::THERMAL:
```
Comment (magicrub) [inline ArduPlane/soaring.cpp:29]: but what happens if we're in Loiter? It just doesn't work now?

Comment (samuelctabor) [reply ArduPlane/soaring.cpp:29]: THERMAL is now what LOITER was with soaring enabled.
LOITER is now what LOITER was without soaring enabled.

File: ArduPlane/control_modes.cpp [hunk 2]
```diff
@@ -73,6 +73,13 @@ Mode *Plane::mode_from_mode_num(const enum Mode::Number num)
     case Mode::Number::TAKEOFF:
         ret = &mode_takeoff;
         break;
+#if SOARING_ENABLED == ENABLED
+    case Mode::Number::THERMAL:
+        ret = &mode_thermal;
+        break;
+#endif
+    default:
```
Comment (peterbarker) [inline ArduPlane/control_modes.cpp:81]: There shouldn't be a `default:` clause here.... put the `#if inside the `case` for thermal.

File: ArduPlane/mode_acro.cpp [hunk 3]
```diff
@@ -1,4 +1,3 @@
-#include "mode.h"
```
Comment (peterbarker) [inline ArduPlane/mode_acro.cpp:1]: This isn't right.

You're now relying on Plane to give you the Mode definitions, which isn't the right way to get them.

Similarly elsewhere.

Getting rid of the Plane.h include would be nice, 'though :-)  Rather unlikely, however, as most will be doing `plane.` something

**Review by peterbarker — DISMISSED** [review]
Will be good for future enhancement, too, as we won't need to worry about how it affects normal loiter.

**Comment by peterbarker** [comment]
Oh - flash cost?

File: ArduPlane/control_modes.cpp (hunk 2 above)
Comment (samuelctabor) [reply ArduPlane/control_modes.cpp]: Thanks, fixed.

**Comment by samuelctabor** [comment]
> Oh - flash cost?


master - 3e882

```
Build directory: /home/samuel/ArdupilotDev/ardupilot/build/Pixhawk1-1M
Target         Text    Data  BSS     Total  
--------------------------------------------
bin/arduplane  965744  1172  195632  1162548
```

```
Build directory: /home/samuel/ArdupilotDev/ardupilot/build/CubeOrange
Target         Text     Data  BSS     Total  
---------------------------------------------
bin/arduplane  1484564  1516  129764  1615844
```

This PR as it stands - 2f59a

```
Build directory: /home/samuel/ArdupilotDev/ardupilot/build/Pixhawk1-1M
Target         Text    Data  BSS     Total  
--------------------------------------------
bin/arduplane  965752  1172  195632  1162556
```

```
Build directory: /home/samuel/ArdupilotDev/ardupilot/build/CubeOrange
Target         Text     Data  BSS     Total  
---------------------------------------------
bin/arduplane  1484832  1516  129760  1616108
```

So 8 bytes with soaring disabled and 264 bytes with soaring enabled.

File: ArduPlane/mode_acro.cpp (hunk 3 above)
Comment (samuelctabor) [reply ArduPlane/mode_acro.cpp:1]: The issue I have is that if `mode.h` is included here, `SOARING_ENABLED` is not yet defined within `mode.h`.

If I include `config.h` in `mode.h`, `AP_AHRS_NAVEKF_AVAILABLE` is not yet defined within `config.h`.

I'll spend some more time on this - any pointers appreciated!

**Comment by samuelctabor** [comment]
I haven't made much progress towards restoring the `mode.h` header inclusion in the flight mode implementations.

This order-of-inclusion problem is common to all vehicles that want to have flight modes optionally included. Only Copter has this right now, and it doesn't include `mode.h` in the implementations.

Perhaps we could discuss this on the EU dev call?

**Comment by samuelctabor** [comment]
Following @andyp1per 's [example](https://github.com/ArduPilot/ardupilot/pull/15405) I moved the soaring define into `AP_Soaring.h` and included this within `mode.h`. Firmware sizes are the same as before.

**Review by peterbarker — APPROVED** [review]

File: libraries/AP_Soaring/AP_Soaring.h [hunk 4]
```diff
@@ -16,6 +16,10 @@
 #include "Variometer.h"
 #include <AP_SpdHgtControl/AP_SpdHgtControl.h>
 
+#ifndef SOARING_ENABLED
```
Comment (tridge) [inline libraries/AP_Soaring/AP_Soaring.h:19]: use HAL_SOARING_ENABLED

Comment (tridge) [reply libraries/AP_Soaring/AP_Soaring.h:19]: also, #if around the SoaringController class

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by tridge** [comment]
need PR against mavlink PLANE_MODE enum
and against pymavlink

**Comment by samuelctabor** [comment]
- Updated to use HAL_SOARING_ENABLED
- Now compiling SoaringController only if HAL_SOARING_ENABLED 
- Opened PRs against [pymavlink](https://github.com/ArduPilot/pymavlink/pull/463) and [mavlink](https://github.com/ArduPilot/mavlink/pull/152)

**Comment by samuelctabor** [comment]
Also updated https://github.com/ArduPilot/ardupilot_wiki/pull/3013 @Hwurzburg.

**Pull request merged**